lightning/src/chain/keysinterface.rs

```diff
@@ -342,6 +340,20 @@ pub trait KeysInterface: Send + Sync {
 	fn get_channel_id(&self) -> [u8; 32];
 }
 
+#[derive(Clone)]
+/// Holds late-bound channel data.
+/// This data is available after the channel is known to be accepted, either
+/// when receiving an open_channel for an inbound channel or when
+/// receiving accept_channel for an outbound channel.
+pub struct AcceptedChannelData {
+	/// Remote public keys and base points
+	pub(crate) remote_channel_pubkeys: ChannelPublicKeys,
+	/// Remote to_self_delay
+	pub(crate) remote_to_self_delay: u16,
+	/// Local to_self_delay
+	pub(crate) local_to_self_delay: u16,
+}
```

Comment:
We need the `local_keys` populated in `Channel` in order to create an `open_channel` message for outbound channels.  They are used to generate the first `per_commitment_point` and the local pubkeys for insertion into that message.

The first two fields in this additional struct will then be available later, when the peer responds with an `accept_channel`.

So I don't think there's a way to create the local keys in one place without breaking some abstraction.

Reply:
Also, a more general point, an external signer does need to be stateful if it wants to keep funds secure against node compromise.  Keeping track of channel state and commitment transaction revocation are part of that needed state.

Reply:
Hum, right, I mean there may be a refactor here that makes sense given non-external-signer users won't do enforcement, but it is probably a bigger deal and unrelated to this PR.

Reply:
The intent of ChannelKeys in this context is that an external signer API (ie API to hardware wallet) needs to initialize the context for a channel (and derive keys for said channel) first, then we can send the OpenChannel message, and then once we receive AcceptChannel we can tell the hardware wallet what our counterparty's public keys are for the channel. We could try to split it up and have a "get the context for the keys you just gave me, and here's my counterparty's keys" API with the local public keys provided via something separate, but that's also somewhat awkward.

Reply:
That doesn't quite help me understand why we need to tell the external signer of all the counterparty's public keys upfront rather than just supplying the appropriate public keys as needed when making calls to it.

Reply:
I started a thread about this in Slack (#hsm channel).

A fully validating external signer would have to check that the node isn't supplying it with remote public keys that vary over time for a channel.  So it seems repetitive to supply data that is checked to be the same and then effectively thrown away.  It would also be less performant if the signer is across a network connection.

Reply:
Performance is one issue, but, at least from my perspective, anything we can do to reduce the number of things an external signer needs to check is one less possible critical vulnerability, so this kind of movement is important. Of course I'm not sure that the to_self_delays could be used for such an attack, having one fewer things to puzzle over is welcome to me.

Reply:
Gotcha. Responded on Slack but may need to wait for @devrandom to come back online before continuing the conversation. I'm largely indifferent as to how the signer gets the necessary data (i.e., upfront vs per call). My concern here is having a signer in a partially initialized state.

To state my concern slightly differently now that I have a better understanding from Slack: "keys" and "signer" are really two different abstractions but are currently conflated into one abstraction. Thus, a signer isn't functional until all the keys are present and so shouldn't be constructed until then, in my view.

@TheBlueMatt You're right that this sort of refactor may possible but unrelated to this PR. I don't mean to unnecessarily hold it up. But I do think we should strive to move to cleaner abstractions as we move forward. It's possible this series of refactorings will get us there. But I have don't have a way of comparing the intended end result against a hypothetical different approach. Maybe @devrandom can open an issue describing the "phase 2" signer that these refactorings are leading to? This would allow for early feedback on the overall design / approach.